File: backend/src/main/java/com/ems/service/EmployeeService.java
package com.ems.service;

import com.ems.entity.Employee;
import com.ems.entity.User;
import com.ems.repository.EmployeeRepository;
import com.ems.repository.UserRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Slf4j
@Service
public class EmployeeService {

    @Autowired
    private EmployeeRepository employeeRepository;

    @Autowired
    private UserRepository userRepository;

    public List<Employee> getAllEmployees() {
        log.info("🔵 Fetching all employees");
        try {
            List<Employee> employees = employeeRepository.findAll();
            log.info("✅ Successfully fetched {} employees", employees.size());
            return employees;
        } catch (Exception e) {
            log.error("❌ Error fetching all employees: {}", e.getMessage(), e);
            throw new RuntimeException("Failed to fetch employees", e);
        }
    }

    public Optional<Employee> getEmployeeById(Long id) {
        log.info("🔵 Fetching employee by ID: {}", id);
        try {
            Optional<Employee> employee = employeeRepository.findById(id);
            if (employee.isPresent()) {
                log.info("✅ Employee found: ID={}, Name={}", id, employee.get().getUser().getFirstName());
            } else {
                log.warn("⚠️ Employee not found: ID={}", id);
            }
            return employee;
        } catch (Exception e) {
            log.error("❌ Error fetching employee by ID {}: {}", id, e.getMessage(), e);
            throw new RuntimeException("Failed to fetch employee", e);
        }
    }

    public Optional<Employee> getEmployeeByEmployeeId(String employeeId) {
        log.info("🔵 Fetching employee by Employee ID: {}", employeeId);
        try {
            Optional<Employee> employee = employeeRepository.findByEmployeeId(employeeId);
            if (employee.isPresent()) {
                log.info("✅ Employee found: EmployeeID={}, Name={}", employeeId, employee.get().getUser().getFirstName());
            } else {
                log.warn("⚠️ Employee not found: EmployeeID={}", employeeId);
            }
            return employee;
        } catch (Exception e) {
            log.error("❌ Error fetching employee by EmployeeID {}: {}", employeeId, e.getMessage(), e);
            throw new RuntimeException("Failed to fetch employee", e);
        }
    }

    public List<Employee> getEmployeesByDepartment(String department) {
        log.info("🔵 Fetching employees by department: {}", department);
        try {
            List<Employee> employees = employeeRepository.findByDepartment(department);
            log.info("✅ Found {} employees in department: {}", employees.size(), department);
            return employees;
        } catch (Exception e) {
            log.error("❌ Error fetching employees by department {}: {}", department, e.getMessage(), e);
            throw new RuntimeException("Failed to fetch employees by department", e);
        }
    }

    public List<Employee> getEmployeesByPosition(String position) {
        log.info("🔵 Fetching employees by position: {}", position);
        try {
            List<Employee> employees = employeeRepository.findByPosition(position);
            log.info("✅ Found {} employees with position: {}", employees.size(), position);
            return employees;
        } catch (Exception e) {
            log.error("❌ Error fetching employees by position {}: {}", position, e.getMessage(), e);
            throw new RuntimeException("Failed to fetch employees by position", e);
        }
    }

    public Employee createEmployee(Employee employee, Long userId) {
        log.info("🔵 Creating new employee. Employee ID: {}, UserId: {}", employee.getEmployeeId(), userId);
        try {
            Optional<User> user = userRepository.findById(userId);
            if (user.isPresent()) {
                employee.setUser(user.get());
                employee.setCreatedAt(LocalDateTime.now());
                employee.setUpdatedAt(LocalDateTime.now());
                Employee savedEmployee = employeeRepository.save(employee);
                log.info("✅ Employee created successfully. ID: {}, Name: {}, Department: {}", 
                    savedEmployee.getId(), savedEmployee.getUser().getFirstName(), savedEmployee.getDepartment());
                return savedEmployee;
            } else {
                log.error("❌ User not found for ID: {}", userId);
                throw new RuntimeException("User not found");
            }
        } catch (Exception e) {
            log.error("❌ Error creating employee: {}", e.getMessage(), e);
            throw new RuntimeException("Failed to create employee", e);
        }
    }

    public Employee updateEmployee(Long id, Employee employeeDetails) {
        log.info("🔵 Updating employee. ID: {}", id);
        try {
            Optional<Employee> employee = employeeRepository.findById(id);
            if (employee.isPresent()) {
                Employee existingEmployee = employee.get();
                if (employeeDetails.getDepartment() != null) {
                    log.debug("Updating department from {} to {}", existingEmployee.getDepartment(), employeeDetails.getDepartment());
                    existingEmployee.setDepartment(employeeDetails.getDepartment());
                }
                if (employeeDetails.getPosition() != null) {
                    log.debug("Updating position from {} to {}", existingEmployee.getPosition(), employeeDetails.getPosition());
                    existingEmployee.setPosition(employeeDetails.getPosition());
                }
                if (employeeDetails.getSalary() != null) {
                    log.debug("Updating salary");
                    existingEmployee.setSalary(employeeDetails.getSalary());
                }
                if (employeeDetails.getPhone() != null) {
                    log.debug("Updating phone");
                    existingEmployee.setPhone(employeeDetails.getPhone());
                }
                if (employeeDetails.getAddress() != null) {
                    log.debug("Updating address");
                    existingEmployee.setAddress(employeeDetails.getAddress());
                }
                existingEmployee.setUpdatedAt(LocalDateTime.now());
                Employee updatedEmployee = employeeRepository.save(existingEmployee);
                log.info("✅ Employee updated successfully. ID: {}, Name: {}", id, updatedEmployee.getUser().getFirstName());
                return updatedEmployee;
            } else {
                log.error("❌ Employee not found for ID: {}", id);
                throw new RuntimeException("Employee not found");
            }
        } catch (Exception e) {
            log.error("❌ Error updating employee ID {}: {}", id, e.getMessage(), e);
            throw new RuntimeException("Failed to update employee", e);
        }
    }

    public void deleteEmployee(Long id) {
        log.info("🔵 Deleting employee. ID: {}", id);
        try {
            Optional<Employee> employee = employeeRepository.findById(id);
            if (employee.isPresent()) {
                Employee emp = employee.get();
                employeeRepository.deleteById(id);
                log.info("✅ Employee deleted successfully. ID: {}, Name: {}", id, emp.getUser().getFirstName());
            } else {
                log.warn("⚠️ Employee not found for deletion. ID: {}", id);
            }
        } catch (Exception e) {
            log.error("❌ Error deleting employee ID {}: {}", id, e.getMessage(), e);
            throw new RuntimeException("Failed to delete employee", e);
        }
    }
}

File: frontend/src/app/services/employee.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { tap } from 'rxjs/operators';
import { LoggerService } from './logger.service';

export interface Employee {
  id: number;
  employeeId: string;
  department: string;
  position: string;
  salary: number;
  hireDate: string;
  phone: string;
  address: string;
  city: string;
  state: string;
  zipCode: string;
  country: string;
  dateOfBirth: string;
}

@Injectable({
  providedIn: 'root'
})
export class EmployeeService {
  private apiUrl = 'http://localhost:8080/api/employees';

  constructor(
    private http: HttpClient,
    private logger: LoggerService
  ) {
    this.logger.info('📦 EmployeeService initialized', null, 'SERVICE');
  }

  getEmployees(): Observable<Employee[]> {
    this.logger.info('🔵 Fetching all employees', null, 'SERVICE');
    return this.http.get<Employee[]>(this.apiUrl).pipe(
      tap(
        (employees: Employee[]) => {
          this.logger.info('✅ Successfully fetched employees', { count: employees.length }, 'SERVICE');
        },
        (error: any) => {
          this.logger.error('❌ Failed to fetch employees', { error: error.message }, 'SERVICE');
        }
      )
    );
  }

  getEmployeeById(id: number): Observable<Employee> {
    this.logger.info('🔵 Fetching employee by ID', { id }, 'SERVICE');
    return this.http.get<Employee>(`${this.apiUrl}/${id}`).pipe(
      tap(
        (employee: Employee) => {
          this.logger.info('✅ Employee fetched successfully', { id, name: `${employee.employeeId}` }, 'SERVICE');
        },
        (error: any) => {
          this.logger.error('❌ Failed to fetch employee', { id, error: error.message }, 'SERVICE');
        }
      )
    );
  }

  createEmployee(employee: Employee, userId: number): Observable<Employee> {
    this.logger.info('🔵 Creating new employee', { userId, department: employee.department }, 'SERVICE');
    return this.http.post<Employee>(`${this.apiUrl}?userId=${userId}`, employee).pipe(
      tap(
        (createdEmployee: Employee) => {
          this.logger.info('✅ Employee created successfully', { id: createdEmployee.id, employeeId: createdEmployee.employeeId }, 'SERVICE');
        },
        (error: any) => {
          this.logger.error('❌ Failed to create employee', { error: error.message }, 'SERVICE');
        }
      )
    );
  }

  updateEmployee(id: number, employee: Employee): Observable<Employee> {
    this.logger.info('🔵 Updating employee', { id, department: employee.department }, 'SERVICE');
    return this.http.put<Employee>(`${this.apiUrl}/${id}`, employee).pipe(
      tap(
        (updatedEmployee: Employee) => {
          this.logger.info('✅ Employee updated successfully', { id }, 'SERVICE');
        },
        (error: any) => {
          this.logger.error('❌ Failed to update employee', { id, error: error.message }, 'SERVICE');
        }
      )
    );
  }

  deleteEmployee(id: number): Observable<void> {
    this.logger.info('🔵 Deleting employee', { id }, 'SERVICE');
    return this.http.delete<void>(`${this.apiUrl}/${id}`).pipe(
      tap(
        () => {
          this.logger.info('✅ Employee deleted successfully', { id }, 'SERVICE');
        },
        (error: any) => {
          this.logger.error('❌ Failed to delete employee', { id, error: error.message }, 'SERVICE');
        }
      )
    );
  }

  getEmployeesByDepartment(department: string): Observable<Employee[]> {
    this.logger.info('🔵 Fetching employees by department', { department }, 'SERVICE');
    return this.http.get<Employee[]>(`${this.apiUrl}/department/${department}`).pipe(
      tap(
        (employees: Employee[]) => {
          this.logger.info('✅ Employees fetched by department', { department, count: employees.length }, 'SERVICE');
        },
        (error: any) => {
          this.logger.error('❌ Failed to fetch employees by department', { department, error: error.message }, 'SERVICE');
        }
      )
    );
  }
}
